src/query/function_score_query.rs

```diff
@@ -0,0 +1,289 @@
+    query: Box<dyn Query>,
+    function: Function,
+}
+
+impl FunctionScoreQuery {
+    /// Creates a new FunctionScoreQuery.
+    pub fn new(
+        query: Box<dyn Query>,
+        function: Function,
+    ) -> Self {
+        FunctionScoreQuery { query, function }
+    }
+}
+
+impl Clone for FunctionScoreQuery {
+    fn clone(&self) -> Self {
+        FunctionScoreQuery {
+            query: self.query.box_clone(),
+            function: self.function.clone(),
+        }
+    }
+}
+
+impl fmt::Debug for FunctionScoreQuery {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "FunctionScore(query={:?})", self.query)
+    }
+}
+
+impl Query for FunctionScoreQuery {
+    fn weight(&self, enable_scoring: EnableScoring<'_>) -> crate::Result<Box<dyn Weight>> {
+        let inner_weight = self.query.weight(enable_scoring)?;
+        Ok(Box::new(FunctionScoreWeight::new(
+            inner_weight,
+            enable_scoring.is_scoring_enabled(),
+            self.function.clone(),
+        )))
+    }
+}
+
+struct FunctionScoreWeight {
+    weight: Box<dyn Weight>,
+    scoring_enabled: bool,
+    function: Function,
+}
+
+impl FunctionScoreWeight {
+    pub fn new(
+        weight: Box<dyn Weight>,
+        scoring_enabled: bool,
+        function: Function,
+    ) -> Self {
+        FunctionScoreWeight {
+            weight,
+            scoring_enabled,
+            function,
+        }
+    }
+}
+
+impl Weight for FunctionScoreWeight {
+    fn scorer(&self, reader: &SegmentReader, boost: Score) -> crate::Result<Box<dyn Scorer>> {
+        let inner_scorer = self.weight.scorer(reader, boost)?;
+        if !self.scoring_enabled {
+            Ok(Box::new(inner_scorer))
+        } else {
+            let new_inner_scorer = self.weight.scorer(reader, boost)?;
+            Ok(Box::new(FunctionScorer::new(
+                inner_scorer,
+                new_inner_scorer,
+                Arc::new(reader.clone()),
+                self.function.clone(),
+            )))
+        }
+    }
+
+    fn explain(&self, reader: &SegmentReader, doc: u32) -> crate::Result<Explanation> {
+        let mut scorer = self.scorer(reader, 1.0)?;
+        if scorer.doc() < doc && scorer.seek(doc) != doc {
+            return Err(TantivyError::InvalidArgument(format!(
+                "Document #({doc}) does not match"
+            )));
+        }
+        if !self.scoring_enabled {
+            return Ok(Explanation::new("FunctionScore with no scoring", 1.0));
+        }
+        let mut explanation = Explanation::new("FunctionScore", scorer.score());
+        let underlying_explanation = self.weight.explain(reader, doc)?;
+        explanation.add_detail(underlying_explanation);
+        Ok(explanation)
+    }
+
+    fn count(&self, reader: &SegmentReader) -> crate::Result<u32> {
+        self.weight.count(reader)
+    }
+}
+
+struct FunctionScorer<TDocSet: DocSet> {
+    docset: TDocSet,
+    scorer: Box<dyn Scorer>,
```

Comment:
Do we really need both? Or rather, I think these must be one scorer to make sure the docset operations and the scoring operations stay synchronized, otherwise `self.docset.doc()` could refer to a different document than `self.scorer.score()`. (I believe the fast field test case should be able to surface this if it is an issue by adding at least one more document.)

Reply:
I have found that actually the `docset` is not really needed so I just deleted it. Now the `FunctionScorer` looks like this:

```rust
struct FunctionScorer {
    scorer: Box<dyn Scorer>,
    segment_reader: Arc<SegmentReader>,
    function: Function,
}
```

Reply:
But could also please extend the fastfield-based test case to actually use this, i.e. have at least two documents which get two different scores so that the earlier issue would have surfaced? Thanks.